network/src/dms2.rs

```diff
@@ -0,0 +1,674 @@
+        );
+        let dms_key_ = dms.read().await.key.clone();
+        if dms_key != dms.read().await.key {
+            return Err(format!("key mismatch: requested {dms_key}, but {dms_key_}"));
+        }
+        for message in messages {
+            let message = message.into_message().map_err(|e| e.to_string())?;
+            dms.write()
+                .await
+                .add_message(message)
+                .await
+                .map_err(|e| e.to_string())?;
+        }
+        Ok(())
+    }
+}
+
+struct DummyFilter;
+
+impl MessageFilter for DummyFilter {
+    fn filter(&self, _message: &Message) -> Result<(), String> {
+        Ok(())
+    }
+}
+
+/// Runs a DMS server. This function will block the current thread.
+pub async fn serve<S: Storage>(
+    dms: Arc<RwLock<DistributedMessageSet<S>>>,
+    network_config: ServerNetworkConfig,
+) -> Result<(), Error> {
+    let port_key = format!("dms-{}", dms.read().await.key);
+    let port = network_config
+        .ports
+        .get(&port_key)
+        .ok_or_else(|| eyre!(format!("`ports` has no field of {port_key}")))?;
+
+    let dms_ = Arc::clone(&dms);
+    let rpc_task = async move {
+        let wrapped_dms = Arc::new(parking_lot::RwLock::new(Some(dms_)));
+        let wrapped_dms_ = Arc::clone(&wrapped_dms);
+        struct DropHelper<T> {
+            wrapped_dms: Arc<parking_lot::RwLock<Option<Arc<RwLock<T>>>>>,
+        }
+        impl<T> Drop for DropHelper<T> {
+            fn drop(&mut self) {
+                self.wrapped_dms.write().take().unwrap();
+            }
+        }
+        let _drop_helper = DropHelper { wrapped_dms };
+        run_server(
+            *port,
+            [(
+                "dms".to_owned(),
+                create_http_object(Arc::new(DmsWrapper { dms: wrapped_dms_ })
+                    as Arc<dyn DistributedMessageSetRpcInterface>),
+            )]
+            .iter()
+            .cloned()
+            .collect(),
+        )
+        .await;
+    };
+    rpc_task.await;
+    Ok(())
+}
+
+/// Runs a DMS client with auto-sync. This function will block the current thread.
+pub async fn sync<S: Storage>(
+    dms: Arc<RwLock<DistributedMessageSet<S>>>,
+    fetch_interval: Option<Duration>,
+    broadcast_interval: Option<Duration>,
+    network_config: ClientNetworkConfig,
+) -> Result<(), Error> {
+    let dms_ = Arc::clone(&dms);
+    let network_config_ = network_config.clone();
+    let fetch_task = async move {
+        if let Some(interval) = fetch_interval {
+            loop {
+                if let Err(e) =
+                    DistributedMessageSet::<S>::fetch(Arc::clone(&dms_), &network_config_).await
+                {
+                    log::warn!("failed to parse message from the RPC-fetch: {}", e);
+                }
+                tokio::time::sleep(interval).await;
+            }
+        } else {
+            futures::future::pending::<()>().await;
```

Comment:
if you reach this line,
won't this task wait indefinitely?

Reply:
Yes, as the comment of the function explained.

Reply:
Then why don't you force `broadcast_interval` to be a value, not option?

Reply:
This function runs indefinitely **regardless of** the value (`Some` or `None`) of both `fetch_interval` and  `broadcast_interval`. That's the interface of this function. It's still stoppable if the user carefully handle the returned future (note that calling an async function itself is a future).

Reply:
I don't understand well why "this function runs indefinitely regardless of" the arguments. Could you please tell me the rationale involved with your decision?